Fix the CAS tests on Windows when LLVM_ENABLE_ONDISK_CAS=OFF

hjyamauchi opened the pull request:
Fix the CAS tests on Windows when LLVM_ENABLE_ONDISK_CAS=OFF
The following tests now pass on Windows:

```
LLVM :: DebugInfo/CAS/AArch64/debug_line_and_info_dedupe.test
LLVM :: DebugInfo/CAS/AArch64/debug_unopt.ll
LLVM :: DebugInfo/CAS/AArch64/multiple_cus.ll
LLVM :: tools/llvm-cas-dump/basic_debug_test.ll
LLVM :: tools/llvm-cas-dump/enums_split.ll
LLVM :: tools/llvm-cas-dump/stats-dump.test
LLVM :: tools/llvm-cas-dump/types_split.ll
LLVM :: tools/llvm-cas-object-format/materialize-objects.s
LLVM-Unit :: CAS/./CASTests.exe/0/28
LLVM-Unit :: CAS/./CASTests.exe/19/28
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/BasicRealFSIteration
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/BasicRealFSRecursiveIteration
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/BasicRealFSRecursiveIterationNoPush
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/BrokenSymlinkRealFSRecursiveIteration
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/ExcludeFromTacking
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/Exists
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/MultipleWorkingDirs
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/TrackNewAccesses
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/TrackNewAccessesExists
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/TrackNewAccessesStack
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/caseSensitivityDir
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/caseSensitivityFile
LLVM-Unit :: Support/./SupportTests.exe/SourceMgrTest/AddIncludedFile
LLVM-Unit :: Support/./SupportTests.exe/TreePathPrefixMapperTest/map
LLVM-Unit :: Support/./SupportTests.exe/TreePathPrefixMapperTest/mapInPlace
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectory/OnDisk
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectory/OnDisk_DisableRemoveOnSignal
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectory/OnDisk_DisableTemporaries
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectoryNested/OnDisk
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectoryNested/OnDisk_DisableRemoveOnSignal
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectoryNested/OnDisk_DisableTemporaries
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectoryNoImply/OnDisk
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectoryNoImply/OnDisk_DisableRemoveOnSignal
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectoryNoImply/OnDisk_DisableTemporaries
```

With LLVM_ENABLE_ONDISK_CAS=OFF, all the CAS tests pass.

Some tests are disabled for now because they work only when
`LVM_ENABLE_ONDISK_CAS=ON` and the lack of support for creating symlinks
in the FileSystem API on windows.

```
    With LLVM_ENABLE_ONDISK_CAS=ON, the following tests currently fail.
    LLVM :: tools/llvm-cas/ingest-blobs-casid.test
    LLVM :: tools/llvm-cas/ingest-remap.test
    LLVM :: tools/llvm-cas/ingest.test
    LLVM :: tools/llvm-cas/merge.test
    LLVM :: tools/llvm-cas/print-id.test
    LLVM-Unit :: CAS/./CASTests.exe/OnDiskCASTest/BlobsBigParallelMultiCAS
    LLVM-Unit :: CAS/./CASTests.exe/OnDiskCASTest/BlobsParallelMultiCAS
    LLVM-Unit :: CAS/./CASTests.exe/PluginCASTest/isMaterialized
```

Will work on fixing the tests with `LVM_ENABLE_ONDISK_CAS=ON` next.

hjyamauchi commented:
This is covers the CAS regression/unit tests in the default build (`ninja check-llvm`) so far. I'll also get to the CAS tests with the  `LLVM_ENABLE_ONDISK_CAS` enabled.

compnerd commented:
CC: @cachemeifyoucan

cachemeifyoucan commented:
@swift-ci please test llvm

cachemeifyoucan commented:
Thanks! If you can fix most of the unit test, that might just be the majority of the work needed for all CAS tests.

The last time I tried OnDiskCAS implementation for Windows in a VM, it doesn't quite work but I didn't have time to debug that. If you can verify all the posix FileSystem APIs are working properly on windows, that would be great.

hjyamauchi commented:
> If you can verify all the posix FileSystem APIs are working properly on windows, that would be great.

@cachemeifyoucan What do you mean exactly? Are there existing tests that verify those?

The CI failed because I failed to cover the clang tests :) I will fix that.

cachemeifyoucan commented:
We don't have a bot to run `check-all` for llvm-project on this repo for Windows. Maybe we should add that.

There are build failures now inside clang. I don't see major problems with this fix other than the build failures need to be addressed.

hjyamauchi commented:
Updated to fix the build issues in clang.

I agree we want to enable the tests in the Windows CI. As there are currently non-CAS-related test failures, we'd need to fix those first, but we can probably start by enabling the llvm unit tests (`check-llvm-unit`) in the CI if we fix one failing one (`SupportTests.exe/rename/ExistingTemp`). I will create a separate PR for that.

hjyamauchi commented:
@swift-ci please test llvm

hjyamauchi commented:
@swift-ci please test llvm

hjyamauchi commented:
@swift-ci please test

hjyamauchi commented:
Updated to fix the swift build errors.

hjyamauchi commented:
@swift-ci please test

hjyamauchi commented:
@swift-ci please test llvm

hjyamauchi commented:
@swift-ci Please test Windows platform

hjyamauchi commented:
It looks like some swift tests are failing on windows. Will look into those.

cachemeifyoucan commented on llvm/include/llvm/CAS/FileSystemCache.h:
```diff
@@ -43,14 +43,15 @@ class FileSystemCache : public ThreadSafeRefCountedBase<FileSystemCache> {
   struct DirectoryListingInfo;
 
   struct LookupPathState {
+    FileSystemCache *Cache;
```
Is all we need here is `PathStyle`? If so, I will avoid pulling in the entire FileSystemCache in case that was mis-used.

hjyamauchi replied:
Done.

cachemeifyoucan reviewed (commented):
LGTM except one comments!

Thanks for fixing everything!

tristanlabelle commented on llvm/include/llvm/Testing/Support/SupportHelpers.h:
```diff
@@ -172,19 +172,23 @@ class TempDir {
 /// The link may be a soft or a hard link, depending on the platform.
```
Given this comment, could we update the implementation to support creating a soft directory link on Windows (probably a directory junction rather than symbolic link), so we don't have to ignore errors?

hjyamauchi replied:
I got the impression that there's a reason not to support soft links on windows in the FileSystem API from the upstream LLVM as symlinks would require a certain system privilege on windows. Based on that, I disabled the test code paths that relies on symlinks for now. I'm not sure if symlinks are necessary for the CAS support at all and/or for this particular PR (if that can be future work after this PR). I updated the PR description to clarify this point. @cachemeifyoucan do you have thoughts on this?

cachemeifyoucan replied:
I think for test coverage, that is fine. We don't need to test soft link on windows for CAS file system.

For CAS File system, the goal is to create an abstract file system representation in CAS that can mimic the real file system. I don't think we need to replicate all aspects of the file system if our tools that is using the CAS file system do not relied on those features. For example, we can easily model symlinks by using two files (since they de-duplicate the CAS anyway) as long as compiler is not trying to check and see if one of them is a symlink or not.

hjyamauchi replied:
Got it 👍

hjyamauchi committed:
Fix the CAS tests on Windows when LLVM_ENABLE_ONDISK_CAS=OFF

The following tests now pass on Windows:

LLVM :: DebugInfo/CAS/AArch64/debug_line_and_info_dedupe.test
LLVM :: DebugInfo/CAS/AArch64/debug_unopt.ll
LLVM :: DebugInfo/CAS/AArch64/multiple_cus.ll
LLVM :: tools/llvm-cas-dump/basic_debug_test.ll
LLVM :: tools/llvm-cas-dump/enums_split.ll
LLVM :: tools/llvm-cas-dump/stats-dump.test
LLVM :: tools/llvm-cas-dump/types_split.ll
LLVM :: tools/llvm-cas-object-format/materialize-objects.s
LLVM-Unit :: CAS/./CASTests.exe/0/28
LLVM-Unit :: CAS/./CASTests.exe/19/28
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/BasicRealFSIteration
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/BasicRealFSRecursiveIteration
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/BasicRealFSRecursiveIterationNoPush
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/BrokenSymlinkRealFSRecursiveIteration
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/ExcludeFromTacking
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/Exists
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/MultipleWorkingDirs
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/TrackNewAccesses
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/TrackNewAccessesExists
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/TrackNewAccessesStack
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/caseSensitivityDir
LLVM-Unit :: CAS/./CASTests.exe/CachingOnDiskFileSystemTest/caseSensitivityFile
LLVM-Unit :: Support/./SupportTests.exe/SourceMgrTest/AddIncludedFile
LLVM-Unit :: Support/./SupportTests.exe/TreePathPrefixMapperTest/map
LLVM-Unit :: Support/./SupportTests.exe/TreePathPrefixMapperTest/mapInPlace
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectory/OnDisk
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectory/OnDisk_DisableRemoveOnSignal
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectory/OnDisk_DisableTemporaries
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectoryNested/OnDisk
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectoryNested/OnDisk_DisableRemoveOnSignal
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectoryNested/OnDisk_DisableTemporaries
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectoryNoImply/OnDisk
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectoryNoImply/OnDisk_DisableRemoveOnSignal
LLVM-Unit :: Support/./SupportTests.exe/VirtualOutput/BackendTest/KeepMissingDirectoryNoImply/OnDisk_DisableTemporaries

With LLVM_ENABLE_ONDISK_CAS=OFF, all the CAS tests pass.

Some tests are disabled for now because they work only when
LVM_ENABLE_ONDISK_CAS=ON and the lack of support for creating symlinks
in the FileSystem API on windows.

With LLVM_ENABLE_ONDISK_CAS=ON, the following tests currently fail.
LLVM :: tools/llvm-cas/ingest-blobs-casid.test
LLVM :: tools/llvm-cas/ingest-remap.test
LLVM :: tools/llvm-cas/ingest.test
LLVM :: tools/llvm-cas/merge.test
LLVM :: tools/llvm-cas/print-id.test
LLVM-Unit :: CAS/./CASTests.exe/OnDiskCASTest/BlobsBigParallelMultiCAS
LLVM-Unit :: CAS/./CASTests.exe/OnDiskCASTest/BlobsParallelMultiCAS
LLVM-Unit :: CAS/./CASTests.exe/PluginCASTest/isMaterialized

Will work on fixing the tests with LVM_ENABLE_ONDISK_CAS=ON next.

hjyamauchi commented:
@cachemeifyoucan Updated to cover the CAS regression tests in addition to the unit tests and the failing swift tests. Because adding `FILE_FLAG_BACKUP_SEMANTICS` to allow `openNativeFile` to open directories on Windows was causing several tests to fail, I decided not to use it and instead use a strategy of checking whether a path is a directory or file first and call `openNativeFile` only for files to avoid the original issue. Also addressed your comment. Please take another look.

hjyamauchi commented:
@swift-ci please test

cachemeifyoucan commented on llvm/test/DebugInfo/CAS/AArch64/debug_line_and_info_dedupe.test:
```diff
@@ -5,6 +5,9 @@ RUN: llc -O0 -cas-friendly-debug-info --filetype=obj --cas-backend --cas=%t/cas
 RUN: llc -O0 -cas-friendly-debug-info --filetype=obj --cas-backend --cas=%t/cas --mccas-casid %t/b.ll -o %t/b.id
 RUN: llvm-cas-dump --die-refs --cas=%t/cas --casid-file %t/a.id %t/b.id | FileCheck %s
 
+// LLVM_ENABLE_ONDISK_CAS is not enabled by default on Windows
```
The ON_DISK_CAS requirement should probably go into `lit.local.cfg` in the same directory so it doesn't need to be added to all the tests.

This requires should be `ondisk_cas`

hjyamauchi replied:
Done

cachemeifyoucan replied:
I was actually taking about something like this: https://github.com/swiftlang/llvm-project/blob/next/llvm/test/MC/CAS/lit.local.cfg

This is equivalent to putting `REQURIES: ondisk_cas` on every file in the same directory. I think it will save lots of explicit requirement in the actual test file.

cachemeifyoucan commented on llvm/test/tools/llvm-cas-dump/basic_debug_test.ll:
```diff
@@ -7,6 +7,9 @@
 
 ; REQUIRES: aarch64-registered-target
 
+; LLVM_ENABLE_ONDISK_CAS is not enabled by default on Windows
```
Same here. We should create a `lit.local.cfg` here.

hjyamauchi replied:
Done

cachemeifyoucan reviewed (commented):
So the current state is that only InMemoryCAS works on Linux but not OnDiskCAS?

hjyamauchi commented:
> So the current state is that only InMemoryCAS works on Linux but not OnDiskCAS?

Yes I believe so.

With LLVM_ENABLE_ONDISK_CAS=OFF, all CAS tests pass.

With LLVM_ENABLE_ONDISK_CAS=ON, the following tests currently fail.

```
  LLVM :: tools/llvm-cas/ingest-blobs-casid.test
  LLVM :: tools/llvm-cas/ingest-remap.test
  LLVM :: tools/llvm-cas/ingest.test
  LLVM :: tools/llvm-cas/merge.test
  LLVM :: tools/llvm-cas/print-id.test
  LLVM-Unit :: CAS/./CASTests.exe/OnDiskCASTest/BlobsBigParallelMultiCAS
  LLVM-Unit :: CAS/./CASTests.exe/OnDiskCASTest/BlobsParallelMultiCAS
  LLVM-Unit :: CAS/./CASTests.exe/PluginCASTest/isMaterialized
```

Will look into these next.

hjyamauchi commented:
@swift-ci please test

hjyamauchi commented:
@cachemeifyoucan please take another look

cachemeifyoucan commented:
> > So the current state is that only InMemoryCAS works on Linux but not OnDiskCAS?
> 
> Yes I believe so.
> 
> With LLVM_ENABLE_ONDISK_CAS=OFF, all CAS tests pass.
> 
> With LLVM_ENABLE_ONDISK_CAS=ON, the following tests currently fail.
> 
> ```
>   LLVM :: tools/llvm-cas/ingest-blobs-casid.test
>   LLVM :: tools/llvm-cas/ingest-remap.test
>   LLVM :: tools/llvm-cas/ingest.test
>   LLVM :: tools/llvm-cas/merge.test
>   LLVM :: tools/llvm-cas/print-id.test
>   LLVM-Unit :: CAS/./CASTests.exe/OnDiskCASTest/BlobsBigParallelMultiCAS
>   LLVM-Unit :: CAS/./CASTests.exe/OnDiskCASTest/BlobsParallelMultiCAS
>   LLVM-Unit :: CAS/./CASTests.exe/PluginCASTest/isMaterialized
> ```
> 
> Will look into these next.

Good to know. I think I once saw the same issues. The current OnDiskCAS implementation makes heavy assumptions of the underlying file system for Posix conformance and windows FS might not full meet the requirements. Maybe we can talk off-line for what is needed for this. @compnerd maybe schedule a meeting for this if you are interested?

Before commit 0046c5d6fff17c3d0b5ab99cfc8c2920922c5474, there was a LazyMappedFileRegion implementation for CAS that I actually tested and worked on windows (but I wasn't sure about the performance). Maybe we can bring that back for windows.

hjyamauchi commented:
This was superseded by https://github.com/swiftlang/llvm-project/pull/11140
